addons/folo/common/src/main/java/org/commonjava/indy/folo/change/FoloTrackingAdjustListener.java

```diff
@@ -0,0 +1,128 @@
+import org.commonjava.indy.folo.model.TrackedContentEntry;
+import org.commonjava.indy.folo.model.TrackingKey;
+import org.commonjava.indy.model.core.StoreKey;
+import org.commonjava.indy.promote.change.PromoteChangeManager;
+import org.commonjava.indy.promote.change.event.PathsPromoteCompleteEvent;
+import org.commonjava.indy.promote.change.event.PromoteCompleteEvent;
+import org.commonjava.indy.promote.model.PathsPromoteRequest;
+import org.commonjava.indy.promote.model.PathsPromoteResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.event.Observes;
+import javax.inject.Inject;
+
+import java.util.Set;
+
+@ApplicationScoped
+public class FoloTrackingAdjustListener
+{
+
+    private final Logger logger = LoggerFactory.getLogger( getClass() );
+
+    @Inject
+    private FoloConfig foloConfig;
+
+    @Inject
+    private FoloRecordCache recordManager;
+
+    @Inject
+    private PromoteChangeManager promoteChangeManager;
+
+    @Inject
+    private StoreDataManager storeManager;
+
+    public void onPromoteComplete( @Observes final PromoteCompleteEvent event )
+    {
+        logger.trace( "Promote COMPLETE: {}", event );
+
+        if ( event instanceof PathsPromoteCompleteEvent )
+        {
+            handlePathsPromoteComplete( ( (PathsPromoteCompleteEvent) event ).getPromoteResult() );
+        }
+    }
+
+    private void handlePathsPromoteComplete( PathsPromoteResult promoteResult )
+    {
+        PathsPromoteRequest req = promoteResult.getRequest();
+        StoreKey source = req.getSource();
+        StoreKey target = req.getTarget();
+
+        TrackingKey trackingKey = getTrackingKey( source );
+        if ( trackingKey == null )
+        {
+            logger.trace( "No tracking key for access to: {}", source );
+            return;
+        }
+
+        // Get the sealed record, client MUST seal the record before promote
+        TrackedContent trackedContent = recordManager.get( trackingKey );
+        if ( trackedContent == null )
+        {
+            logger.trace( "No sealed record found, trackingKey: {}", trackingKey );
+            return;
+        }
+
+        String error = promoteResult.getError();
+        if ( error != null )
+        {
+            logger.trace( "Error in promoteResult, skip adjust" );
+            return;
+        }
+
+        Set<String> paths = promoteResult.getCompletedPaths();
+
+        if ( paths.isEmpty() )
+        {
+            logger.trace( "No completedPaths, skip adjust" );
+            return;
+        }
+
+        adjustTrackedContent( trackedContent, source, target );
+
+        recordManager.replaceTrackingRecord( trackedContent );
+    }
+
+    private void adjustTrackedContent( TrackedContent trackedContent, StoreKey source, StoreKey target )
+    {
+        Set<TrackedContentEntry> uploads = trackedContent.getUploads();
+        uploads.forEach( entry ->
```

Comment:
downloads might be affected here too. See my comments above about stuff downloaded from maven central.